python/lsst/ctrl/mpexec/preExecInit.py

```diff
@@ -221,3 +233,36 @@ def logConfigMismatch(msg):
                 # butler will raise exception if dataset is already there
                 _LOG.debug("Saving Config for task=%s dataset type=%s", taskDef.label, configName)
                 self.butler.put(taskDef.config, configName, {})
+
+    def savePackageVersions(self, graph):
+        """Write versions of software packages to butler.
+
+        Parameters
+        ----------
+        graph : `~lsst.pipe.base.QuantumGraph`
+            Execution graph.
+
+        Raises
+        ------
+        Exception
+            Raised if ``checkExisting`` is ``True`` but versions are not
+            compatible.
+        """
+        packages = Packages.fromSystem()
+        datasetType = "packages"
+        oldPackages = self.butler.get(datasetType, {}) if self.skipExisting else None
+        if oldPackages is not None:
+            # Note that because we can only detect python modules that have been imported, the stored
+            # list of products may be more or less complete than what we have now.  What's important is
+            # that the products that are in common have the same version.
```

Comment:
I think the set of imported packages will be sufficiently complete after we've loaded all `PipelineTask` config instances, and probably not complete before that.  Do you happen to know if the sequencing of operations is such that this will be called after that happens?

Reply:
I though about that and figured that it should work because the complete graph is in memory (with some caveat) and it means that config instances are loaded too which should import all relevant code.
The caveat is that BPS is executing this not for the whole graph but one task at a time so there is only one task configuration loaded and not whole graph. But this still should work because we update versions incrementally if they don't conflict. And I think BPS will probably do --init-only on the whole graph in the future.